scrypt.go

```diff
@@ -212,3 +213,70 @@ func Cost(hash []byte) (Params, error) {
 
 	return params, err
 }
+
+func (p Params) printKey(salt, dk []byte) []byte {
+	return []byte(fmt.Sprintf("%d$%d$%d$%x$%x", p.N, p.R, p.P, salt, dk))
+}
+
+func (p Params) generatePasswordWithLimit(password []byte, timeout time.Duration, memBytes int) ([]byte, Params, error) {
+	if p.N == 0 || p.R == 0 || p.P == 0 || p.SaltLen == 0 || p.DKLen == 0 {
+		p = DefaultParams
+	} else if err := p.Check(); err != nil {
+		return nil, p, err
+	}
+
+	deadline := time.Now().Add(timeout)
+	salt, err := GenerateRandomBytes(p.SaltLen)
+	if err != nil {
+		return nil, p, err
+	}
+
+	var dur time.Duration
+	var dk []byte
+	lng := p
+	p.N <<= 1
+	now := time.Now()
+	// memory = (128 * r * N).
```

Comment:
We should clarify this - `Memory usage is approximately 128 * r * N` - since it is approximate only.